Update wording, links in key projects page (#691)
pypa/packaging.python.org: source/key_projects.rst

`Bitbucket <https://bitbucket.org/pypa/distlib>`__ |
`PyPI <https://pypi.org/project/distlib>`__

Distlib is a library which implements low-level functions that relate
to packaging and distribution of Python software.  It consists in part
of the functions from the `distutils2
<https://pypi.org/project/Distutils2>`_ project, which was intended to

> be released as ``packaging`` in the Python 3.3 stdlib, but was removed

Author: Thanks! Fixed, I believe..

shortly before Python 3.3 entered beta testing. distlib implements
several relevant PEPs (Python Enhancement Proposal standards) and is
useful for developers of third-party packaging tools to make and
upload binary and source distributions, achieve interoperability,
resolve dependencies, manage package resources, and do other similar

> functions.

Author: Thanks - added.

`Docs <https://packaging.pypa.io>`__ |
`Dev list <http://groups.google.com/group/pypa-dev>`__ |
`Issues <https://github.com/pypa/packaging/issues>`__ |
`GitHub <https://github.com/pypa/packaging>`__ |
`PyPI <https://pypi.org/project/packaging>`__ |
User irc:#pypa |
Dev irc:#pypa-dev

Core utilities for Python packaging used by :ref:`pip` and :ref:`setuptools`.

The core utilities in the packaging bundle handle version handling,
specifiers, markers, requirements, tags, and similar attributes and
tasks for Python packages. Most Python users rely on this bundle
without needing to explicitly call it; developers of the other Python
packaging, distribution, and installation tools listed here often use
its functionality to parse, discover, and otherwise handle dependency

> attributes.

Reviewer: ```suggestion
attributes.

This project specifically focuses on implementing the modern Python packaging
interoperability standards defined at :ref:`packaging-specifications`, and will report
errors for sufficiently old legacy packages that are incompatible with those
standards. The :ref:`distlib` project is a more permissive library that attempts to
provide a plausible reading of ambiguous metadata in cases where ``packaging``
will instead report on error.
```

It seems worth highlighting that the essential difference between "packaging" and "distlib" is that the latter includes extra behaviour that isn't in the approved PEPs, and one of the key aspects of that is making more guesses when confronted with ambiguous legacy metadata.

Author: Thanks so much! Added.

`Docs <https://packaging.pypa.io>`__ |
`Dev list <http://groups.google.com/group/pypa-dev>`__ |
`Issues <https://github.com/pypa/packaging/issues>`__ |
`GitHub <https://github.com/pypa/packaging>`__ |
`PyPI <https://pypi.org/project/packaging>`__ |
User irc:#pypa |
Dev irc:#pypa-dev

Core utilities for Python packaging used by :ref:`pip` and :ref:`setuptools`.

> The core utilities in the packaging bundle handle version handling,

Reviewer 2: The word "bundle" surprised me; I hadn't seen it used in the context of Python packaging. Wondering if "library" is more accurate?

```suggestion
The core utilities in the packaging library handle version handling,
```

specifiers, markers, requirements, tags, and similar attributes and

> tasks for Python packages. Most Python users rely on this bundle

Reviewer 2: ```suggestion
tasks for Python packages. Most Python users rely on this library
```

without needing to explicitly call it; developers of the other Python
packaging, distribution, and installation tools listed here often use
its functionality to parse, discover, and otherwise handle dependency
attributes.

This project specifically focuses on implementing the modern Python
packaging interoperability standards defined at
:ref:`packaging-specifications`, and will report errors for
sufficiently old legacy packages that are incompatible with those
standards. In contrast, the :ref:`distlib` project is a more
permissive library that attempts to provide a plausible reading of

> ambiguous metadata in cases where ``packaging`` will instead report on

Reviewer 2: Not sure what the convention is (or should be) around monospacing package names. It seems like most package names are *not* monospaced, except here, once in "Distlib", and in "distutils".

Author: Yeah, I think there is no consistent convention. Future person who decides on a style guide, consider this question!

`User list <http://groups.google.com/group/python-virtualenv>`__ [1]_ |
`Dev list <http://groups.google.com/group/pypa-dev>`__ |
`Issues <https://github.com/pypa/pip/issues>`__ |
`GitHub <https://github.com/pypa/pip>`__ |
`PyPI <https://pypi.org/project/pip/>`__ |
User irc:#pypa |
Dev irc:#pypa-dev

The most popular tool for installing Python packages, and the one

> included with modern versions of Python.

Reviewer: ```suggestion
included with modern versions of Python.

It provides the essential core features for finding, downloading, and installing packages
from pypi.org and other index servers, and can be incorporated into a wide range of
development workflows via its CLI interface.
```

Author: Thank you! Added.

`User list <http://groups.google.com/group/python-virtualenv>`__ [1]_ |
`Dev list <http://groups.google.com/group/pypa-dev>`__ |
`Issues <https://github.com/pypa/pip/issues>`__ |
`GitHub <https://github.com/pypa/pip>`__ |
`PyPI <https://pypi.org/project/pip/>`__ |
User irc:#pypa |
Dev irc:#pypa-dev

The most popular tool for installing Python packages, and the one
included with modern versions of Python.

It provides the essential core features for finding, downloading, and

> installing packages from pypi.org and other index servers, and can be

Reviewer 2: This feels more in line with other references to these concepts:

```suggestion
installing packages from PyPI and other Python package indexes, and can be
```

`Mailing list <http://mail.python.org/mailman/listinfo/distutils-sig>`__ [2]_ |
`Issues <https://github.com/pypa/twine/issues>`__ |
`GitHub <https://github.com/pypa/twine>`__ |
`PyPI <https://pypi.org/project/twine>`__

> Twine is a utility for interacting with PyPI, that offers a secure replacement for

Reviewer: We should update this description, as ``setup.py upload`` is secure in modern versions of Python (since 3.5.3 or so).

``twine`` is still better - just for reasons other than this one.

Author: I updated it more thoroughly. Thanks.

`Mailing list <http://mail.python.org/mailman/listinfo/distutils-sig>`__ [2]_ |
`Issues <https://github.com/pypa/twine/issues>`__ |
`GitHub <https://github.com/pypa/twine>`__ |
`PyPI <https://pypi.org/project/twine>`__

Twine is the primary tool developers use to upload packages to the

> Python Package Index or other Python package indexes. It is a

Reviewer 2: ```suggestion
PyPI or other Python package indexes. It is a
```

> Python Package Index or other Python package indexes. It is a

Reviewer 2: How about ths, to avoid the repetition of "package index"?

```
Twine is the primary tool developers use to upload packages to
PyPI or other Python package indexes. It is a
```

It looks like the acronym is more common in this document than the words.

Author: It is. I'm guessing a lot of people come to section anchor tags in this document and skim around, and I think it's okay to say the full name once in a while. I'm not going to figure out a consistent style guide and apply it in this particular PR, but if someone else does so in a later change and it changes my choices in this regard, then I will be fine with that; I will not cry out to the heavens and rend my clothes or anything,

`User list <http://groups.google.com/group/python-virtualenv>`__ |
`Dev list <http://groups.google.com/group/pypa-dev>`__ |
`Issues <https://github.com/pypa/virtualenv/issues>`__ |
`GitHub <https://github.com/pypa/virtualenv>`__ |
`PyPI <https://pypi.org/project/virtualenv/>`__ |
User irc:#pypa  |
Dev irc:#pypa-dev

virtualenv is a tool which uses the command-line path environment
variable to (like venv) create isolated Python :term:`Virtual

> Environments <Virtual Environment>`. virtualenv goes beyond venv by

Reviewer 3: ```
virtualenv is a tool to create isolated Python :term:`Virtual
Environments <Virtual Environment>` (like venv). virtualenv provides
additional functionality compared to venv by...
```

Author: Thanks. Fixed.

`Docs <https://wheel.readthedocs.io/en/latest/>`__ |
`Mailing list <http://mail.python.org/mailman/listinfo/distutils-sig>`__ [2]_ |
`Issues <https://github.com/pypa/wheel/issues>`__ |
`GitHub <https://github.com/pypa/wheel>`__ |
`PyPI <https://pypi.org/project/wheel>`__ |
User irc:#pypa  |
Dev irc:#pypa-dev

Primarily, the wheel project offers the ``bdist_wheel`` :ref:`setuptools` extension for
creating :term:`wheel distributions <Wheel>`.  Additionally, it offers its own
command line utility for creating and installing wheels.

> See also `auditwheel <https://github.com/pypa/auditwheel>`__, a tool

Reviewer 3: Any specific reason to not make auditwheel a dedicated heading?

Author: Two reasons.

1. `auditwheel` is, as I understand it, only useful in the context of `wheel`.
2. Alphabetically, they're super far from each other.

But I don't particularly care. If you would like to block this PR on that, please say so; otherwise, I figure if someone would like to do that, they can do so and I will not object. :-)

Reviewer 4: This is questionable: technically, it's useful for making manylinux wheels but `wheel` dist isn't used directly. You could create a wheel using any tool you like and then use `auditwheel` to inspect it and/or make it manylinux compliant. Most of the guides just use `pip wheel` command to make that first step. So whether it's only useful in the context of `wheel` probably depends on the PoV.

Reviewer: I was going to comment on this in my initial review, but came to the same conclusion as @brainwane: with the current layout of this page, giving auditwheel its own section would look quite odd, whereas it fits better here.

If we were to give it a top level section, I'd suggest doing it in a separate PR where we categorise this list by project type (e.g. installation, build & publishing, package hosting, utility library) rather than splitting it out here.

Reviewer 3: > If you would like to block this PR on that, please say so

Nope. :)

====

> `Docs <https://github.com/pipxproject/pipx>`__ |

Reviewer 2: ```suggestion
`Docs <https://pipxproject.github.io/pipx/>`__ |
```

Reviewer 4: This suggested change breaks indentation :point_up: (which is a known bug @ GH)
Here's a fixed suggestion:
```suggestion
 `Docs <https://pipxproject.github.io/pipx/>`__ |
```

`GitHub <https://github.com/pipxproject/pipx>`__ |
`PyPI <https://pypi.org/project/pipx/>`__

pipx is a tool to safely install and run Python CLI applications globally.

.. _scikit-build:

piwheels
========

`Website <https://piwheels.org/>`__ |
`Docs <https://piwheels.readthedocs.io/>`__ |
`GitHub <https://github.com/piwheels/piwheels/>`__

Piwheels is a website, and software underpinning it, that fetches

> source code distribution packages from the Python Package Index and

Reviewer 2: ```suggestion
source code distribution packages from PyPI and
```

Similar to my comment re: Twine

compiles them into binary wheels that are optimized for installation
onto Raspberry Pi computers. Many Raspberry Pi users who use or
develop Python tools to run on Raspbian use piwheels as their primary

> Python package index.

Reviewer 4: Recently, I've discovered that this index doesn't support PEP 503 which makes Pip explode for wheels that specify `Requires-Python` under interpreter versions that don't match the condition instead of selecting wheels properly on the download stage. (https://github.com/cherrypy/cherrypy/issues/1837)

I think this may be a good place to add a known issue warning about this.

Author: I will do so; I'll link to https://github.com/piwheels/piwheels/issues/208 which is where they are fixing it. Thanks for the heads-up.

.. _scikit-build:

pip-tools
=========

`GitHub and Docs <https://github.com/jazzband/pip-tools/>`__ |
`PyPI <https://pypi.org/project/pip-tools/>`__

pip-tools is a suite of tools meant for Python system administrators
and release managers who particularly want to keep their builds
deterministic yet stay up to date with new versions of their
dependencies. Users can specify particular release of their
dependencies via hash, conveniently make a properly formatted list of
requirements from information in other parts of their program, update
all dependencies (a feature pip currently does not provide), and
create layers of constraints for the program to obey.

piwheels
========

`Website <https://piwheels.org/>`__ |
`Docs <https://piwheels.readthedocs.io/>`__ |
`GitHub <https://github.com/piwheels/piwheels/>`__

Piwheels is a website, and software underpinning it, that fetches
source code distribution packages from PyPI and compiles them into
binary wheels that are optimized for installation onto Raspberry Pi
computers. Many Raspberry Pi users who use or develop Python tools to
run on Raspbian use Piwheels as their primary Python package index.

> (Note that Piwheels `does not yet fully support

Reviewer 4: This should probably be a warning box marked in RST...

Reviewer 4: ```suggestion
  .. warning::

    Note that Piwheels `does not yet fully support
```

<https://github.com/piwheels/piwheels/issues/208>`__ :pep:`503` and

> thus some wheels have trouble with inconsistent ``Requires-Python``

Reviewer 4: This sounds wrong. Wheels don't have troubles, they are correct, it's Pip that has troubles because it does not get enough info from the index and cannot properly identify what to start downloading.

Reviewer 4: I'd maybe go as far as pointing to pip-tools as a workaround mentioning that this will help end-users pin versions of transitive deps.

Author: At this point I'm going to write a very simplified line of explanation here and I suggest you make a separate PR or commit with the language that you prefer. :-)

User irc:#pypa  |
Dev irc:#pypa-dev

> 

Reviewer 4: Technically it's a package https://github.com/python/cpython/tree/master/Lib/distutils
(module == one py file, package == dir)

Also, this change is inconsistent with the paragraph about ensurepip above.

Author: Thanks. Resolved.

> A module in the Python Standard Library that has support for creating and

Reviewer: ```suggestion
The original Python packaging system, added to the standard library in Python 2.0.
Due to the challenges of maintaining a packaging system where feature updates are tightly coupled to language runtime updates, direct usage of ``distutils`` is now actively discouraged, with :ref:`Setuptools` being the preferred replacement. :ref:`Setuptools` not only provides features that plain ``distutils`` doesn't offer (such as dependency declarations and entry point declarations), it also provides a consistent build interface and feature set across all supported Python versions.
```

Author: Thanks so much! Incorporated your addition.

used immediately after cloning from github.

zest.releaser
=============

`Docs <https://zestreleaser.readthedocs.io/en/latest/>`__ |
`GitHub <https://github.com/zestsoftware/zest.releaser/>`__ |
`PyPI <https://pypi.org/project/zest.releaser/>`__

> zest.releaser is a Python package release tool providing an

Reviewer 4: How about this?
```suggestion
``zest.releaser`` is a Python package release tool providing an
```
It seems like there's no consistency for marking dist names with `<code>` on this page. You added it for distutils so maybe others deserve this too.
